core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -168,6 +178,13 @@ private[spark] abstract class YarnSchedulerBackend(
         logInfo(s"ApplicationMaster registered as $am")
         amEndpoint = Option(am)
 
+      case ReregisterClusterManager(am) =>
```

Comment:
> Is it robust enough to use amEndpoint to distinguish whether AM is registered again or not?

That's the same that would happen if you have a 3rd attempt of running the AM. So better to handle registration the same way for all attempts.